src/_igraph/graphobject.c

```diff
@@ -7740,6 +7740,104 @@ PyObject *igraphmodule_Graph_layout_sugiyama(
 
   igraph_vector_int_destroy(&extd_to_orig_eids);
   return (PyObject *) result_o;
+
+
+  /** \ingroup python_interface_graph
+ * \brief Places the vertices of a graph using Uniform Manifold Approximation and Projection (UMAP)
+ * \return the calculated coordinates as a Python list of lists
+ * \sa igraph_layout_umap
+ */
+PyObject *igraphmodule_Graph_layout_umap(igraphmodule_GraphObject * self,
+                                        PyObject * args, PyObject * kwds)
+{
+  static char *kwlist[] =
+    { "dist", "dim", "min_dist", "epochs", "sampling_prob", NULL };
+  igraph_matrix_t m;
+  igraph_matrix_t *dist = 0;
+  /* FIXME: use reasonable defaults */
+  double min_dist = 0.01, sampling_prob = 0.5;
+  Py_ssize_t dim = 2, Py_ssize_t epochs = 50;
+  PyObject *dist_o = Py_None;
+  PyObject *result_o;
+
+  /* arpack_options_o is now unused but we kept here for sake of backwards
+   * compatibility */
+
+  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|Ondnd", kwlist, &dist_o,
+                                   &dim, &min_dist, &epochs, &sampling_prob))
+    return NULL;
+
+  CHECK_SSIZE_T_RANGE_POSITIVE(dim, "number of dimensions");
+  CHECK_SSIZE_T_RANGE_POSITIVE(epochs, "number of epochs");
+
+  if (min_dist < 0) {
+    /* FIXME: How to actually tell the user? */
+    igraphmodule_handle_igraph_error();
+  }
+
+  if (sampling_prob <= 0) {
+    /* FIXME: How to actually tell the user? */
+    igraphmodule_handle_igraph_error();
+  }
+
+  if (sampling_prob > 1) {
+    /* FIXME: How to actually tell the user? */
+    igraphmodule_handle_igraph_error();
+  }
+
+  if (dist_o != Py_None) {
+    dist = (igraph_matrix_t*)malloc(sizeof(igraph_matrix_t));
```

Comment:
What is the purpose of allocating this dynamically? @ntamas I can see this pattern throughout this file, but isn't it just complicating the code and providing one more chance to forget `free`? We just fixed a bunch of leaks like this.

Reply:
Good question... @ntamas?

Reply:
Yes, that's unnecessary, you could allocate it on the stack. I was inexperienced when I initially wrote code like this, and then it got copied all over the place.

Reply:
no problem. Would you like me to fix these here throughout?

Reply:
one thing at a time :) we can do it outside the scope of this PR. Feel free to add an issue for it and then I'll do it eventually - or someone else can, but let's not mix more things in here

Reply:
FWIW I checked the code more carefully and it seems like this pattern is used for optional input arguments; for instance, in the interface for `layout_mds()`, the user _may_ provide a distance matrix but it is not necessary. Currently the pattern is that `dist == 0` if the user did not provide a matrix, and `dist != 0` means that the user did provide one in Python, in which case we allocate a corresponding C-level matrix on the heap. We could pre-allocate one on the stack and then we can get rid of the `free()` call, but then we need a `p_dist` pointer that is either `NULL` (if the user did not provide a distance matrix in Python) or points to the pre-allocated matrix if needed.

There are maybe 4 or 5 similar occasions in `graphobject.c` so I don't think it's a big deal. If one of you decides to simplify it by using a pre-allocated matrix, please do it in a separate PR so it is reviewed more easily.